### PR TITLE
Added in CreatureTypeSpawnEvent Closes #3808

### LINKED ISSUE #3803: Client Side events
I'm looking for an event that will be seen client side when the first tick starts, or when a game starts. FMLServerStartingEvent works for singleplayer but what about when connecting to a server?

### LINKED ISSUE #3808: [New Event Suggestion] EntitySpawnTypeCheckEvent
Currently with the entity spawning system you are unable to spawn entities in lava or a custom fluid without writing your own entity spawning code. I am suggesting that an EntitySpawnTypeCheckEvent be added. It would be called in WorldEntitySpawner#canCreatureTypeSpawnAtLocation and it would allow entities to spawn in lava and custom fluids. MCF Forum link for more info: http://www.minecraftforge.net/forum/topic/47769-1112-spawn-mob-in-lava/

### PR DESCRIPTION
**Pull request by LogicTechCorp**
With the current entity spawning system, entities are unable to spawn in a fluid that is not water without custom entity spawning code. This PR fixes that issue by adding in an event that allows modders to control if a certain EntityLiving.SpawnPlacementType can spawn at a given location.

### CODE REVIEW

**Comment by Madsthunder** [comment]
I think you meant to refer to #3808 in the title

File: patches/minecraft/net/minecraft/world/WorldEntitySpawner.java.patch [hunk 1]
```diff
@@ -50,7 +50,19 @@
...
-@@ -213,8 +217,9 @@
+@@ -206,6 +210,11 @@
+         {
+             IBlockState iblockstate = p_180267_1_.func_180495_p(p_180267_2_);
+ 
++            net.minecraftforge.fml.common.eventhandler.Event.Result canSpawn = net.minecraftforge.event.ForgeEventFactory.canCreatureTypeSpawnAtPos(p_180267_0_, p_180267_1_, p_180267_2_, iblockstate);
++
++            if (canSpawn == net.minecraftforge.fml.common.eventhandler.Event.Result.ALLOW) return true;
++            else if (canSpawn == net.minecraftforge.fml.common.eventhandler.Event.Result.DENY) return false;
++
```
Comment (mezz) [inline patches/minecraft/net/minecraft/world/WorldEntitySpawner.java.patch]: Please reduce the lines changed here by removing the newlines you have added.

**Review by mezz — CHANGES_REQUESTED** [review]

**Comment by LogicTechCorp** [comment]
Are there any other changes you would like me to make @mezz?

**Comment by mezz** [comment]
Tested and looks good. 
The logic of the test mod isn't great, for a real mod you probably don't want to use `ALLOW` in most cases, just `DENY` or `DEFAULT`. It does work the way it's written though.

**Review by mezz — APPROVED** [review]

**Comment by mezz** [comment]
Reviewing with Lex:
Can you make this into a method on Block instead of an event? Then the block can decide if creatures can spawn there.

**Comment by LogicTechCorp** [comment]
Can you take a look at this MinecraftForge forum thread? It explains the reasoning for the event. @mezz
http://www.minecraftforge.net/forum/topic/47769-1112-spawn-mob-in-lava/

**Comment by LogicTechCorp** [comment]
Is the reasoning behind the event sound or would you still like me to make it a method in the block class?
 @mezz

**Comment by mezz** [comment]
How about a method on the Entity? This event gives a lot of control, but if you're just implementing your own block or entity, it would be simpler and more performant to just create methods on those.

**Comment by LexManos** [comment]
Both entity and block veto would be the best option. We do not want to bypass the block.canCreatureSpawn call, we just want to expand it to also include IN_WATER types?
Or are there other types that are being added that is the issue?

Its a interesting thing, maybe making the default of canCreatureSpawn be true, and only applying the solid logic for type IN_AIR/ON_GROUND like vanilla? Needs more thought.  An event however will not be accepted.

**Comment by LogicTechCorp** [comment]
I think adding in a method to EntityLiving would be the best solution now that I think about it. The only issue I can see with this is that when `canCreatureTypeSpawnHere` is called the entity has not yet been created. What would you suggest for me to do? @mezz

**Comment by mezz** [comment]
Yeah, unfortunately the entity is only available just after a successful "can spawn here" check.

I'll ask Lex to take another look at this after you answer his question about the intention of this PR:
>We do not want to bypass the block.canCreatureSpawn call, we just want to expand it to also include IN_WATER types?
Or are there other types that are being added that is the issue?

**Comment by LogicTechCorp** [comment]
@mezz Oh, sorry about that. My intention is to not bypass `block.canCreatureSpawn`, but rather allow for the entity to also decide if they can spawn at said location.

@LexManos It would be other types that are being added that is the issue, but let me give you an example in case I understood the question wrong. 

I have an entity that should spawn in a fluid that is not water. There are no `EntityLiving.SpawnPlacementType`'s that cover this case. If I were to add in my own `EntityLiving.SpawnPlacementType` via `EnumHelper` it would not be called anywhere.

**Comment by LexManos** [comment]
It WOULD be called somewhere, you have to register it in EntitySpawnPlacementRegistry. Or just pass it to canCreatureTypeSpawnAtLocation.
This PR has nothing to do with making custom SpawnPlacementType's 'called'. 
Seems to me like a better option would be a 'isValidSpawn(world, pos)' added to SpawnPlacementType {Or a map of Type->BiPredicate} 
Combine with making the default return of canCreatureSpawn {or a new function to keep compatibility} be true and only applying logic for GROUND/WATER. 

Again, as I stated before. An event for this WILL NOT be accepted. This can be done much better/more efficiently just asking the block/type.

**Comment by LogicTechCorp** [comment]
Thanks for the feedback, I will look into a better way of achieving this.

**Pull request closed without merging**